### PR TITLE
feat: options dialog and template application 

### PR DESCRIPTION
**Pull request by Cerchie**
## Summary of Changes

Background:

This PR follows 2 others:
- a [PR](https://github.com/confluentinc/intellij/pull/432) to filter templates by IDE 
- - template [selection action](https://github.com/confluentinc/intellij/pull/424)

It's the last step in implementing project generation MVP, and will be followed by PRs to surface it to more obvious areas in the interface. 

### Click-testing instructions

1. Cmd+shift+a, "Select Project Template", then select a template. 

3. Make sure that each option is verified correctly
Screenshot 2026-04-23 at 2 02 22 PM

4. Make sure that closing the form exits the flow properly, no side effects.  

6. Make sure that the Flink UDFs template, which has no options, takes you straight to selecting the folder where the new project will be generated and that the new project is generated properly. 
Screenshot 2026-04-23 at 2 03 46 PM

7. Edge case: verify that leaving options empty renders a template with empty strings in the `.env` file

## Pull request checklist

Please check if your PR fulfills the following (if applicable):

#### Tests

- [x] Added new
- [ ] Updated existing
- [ ] Deleted existing

### CODE REVIEW

File: resources/messages/KafkaBundle.properties [hunk 1]
```diff
@@ -1158,3 +1158,9 @@ scaffold.action.error.title=Project Templates Error
 scaffold.action.error.message=Failed to load project templates: {0}
 scaffold.dialog.template.unknown=Unknown Template
 scaffold.action.fetch.templates.progress=Loading project templates\u2026
+scaffold.options.dialog.title=Configure Template Options
+scaffold.options.validation.min.length={0} must be at least {1} characters
+scaffold.options.validation.pattern={0} does not match the required format
+scaffold.action.apply.template.progress=Generating project from template\u2026
```
Comment (Cerchie) [inline resources/messages/KafkaBundle.properties:1164]: `\u2026` is an ellipsis, see 1160 for precendent

File: src/io/confluent/intellijplugin/scaffold/actions/SelectScaffoldTemplateAction.kt [hunk 2]
```diff
@@ -2,31 +2,62 @@ package io.confluent.intellijplugin.scaffold.actions
...
+        ::ScaffoldTemplateOptionsDialog,
     private val templateSorter: (List<ScaffoldV1TemplateListDataInner>) -> List<ScaffoldV1TemplateListDataInner> =
-        { templates -> IdeLanguageMapper.sortByPreferredLanguage(templates) }
+        { templates -> IdeLanguageMapper.sortByPreferredLanguage(templates) },
+    private val fileChooser: (Project) -> VirtualFile? = { project ->
+        FileChooser.chooseFile(
+            FileChooserDescriptorFactory.createSingleFolderDescriptor()
+                .withTitle(KafkaMessagesBundle.message("scaffold.action.choose.location.title")),
+            project,
+            null
```
Comment (Cerchie) [inline src/io/confluent/intellijplugin/scaffold/actions/SelectScaffoldTemplateAction.kt:49]: This `null` param means the chooser will fall back to its default location, e.g., last used dir / project dir / IDE default

File: src/io/confluent/intellijplugin/scaffold/actions/SelectScaffoldTemplateAction.kt [hunk 3]
```diff
@@ -52,15 +83,65 @@ class SelectScaffoldTemplateAction(
                 templateSorter(fetchTemplates())
             }
 
-            withContext(Dispatchers.EDT + ModalityState.defaultModalityState().asContextElement()) {
-                if (project.isDisposed) return@withContext
+            val userInput = withContext(Dispatchers.EDT + ModalityState.defaultModalityState().asContextElement()) {
+                if (project.isDisposed) return@withContext null
```
Comment (Cerchie) [inline src/io/confluent/intellijplugin/scaffold/actions/SelectScaffoldTemplateAction.kt:87]: stops background processes if project is closed

File: src/io/confluent/intellijplugin/scaffold/actions/SelectScaffoldTemplateAction.kt [hunk 4]
```diff
@@ -52,15 +83,65 @@ class SelectScaffoldTemplateAction(
...
 
                 val dialog = dialogFactory(project, sortedTemplates)
-                if (dialog.showAndGet()) {
-                    val selected = dialog.selectedTemplate
-                    if (selected != null) {
-                        thisLogger().debug("Selected template: ${selected.spec.displayName ?: selected.spec.name}")
-                    }
+                if (!dialog.showAndGet()) return@withContext null
```
Comment (Cerchie) [inline src/io/confluent/intellijplugin/scaffold/actions/SelectScaffoldTemplateAction.kt:90]: also if the user closes the project; aborts action before it opens the options dialog

File: src/io/confluent/intellijplugin/scaffold/actions/SelectScaffoldTemplateAction.kt [hunk 5]
```diff
@@ -52,15 +83,65 @@ class SelectScaffoldTemplateAction(
...
+
+                val selected = dialog.selectedTemplate ?: return@withContext null
+                thisLogger().debug("Selected template: ${selected.spec.displayName ?: selected.spec.name}")
+
+                val options = if (!selected.spec.options.isNullOrEmpty()) {
+                    val optionsDialog = optionsDialogFactory(project, selected)
+                    if (!optionsDialog.showAndGet()) return@withContext null
```
Comment (Cerchie) [inline src/io/confluent/intellijplugin/scaffold/actions/SelectScaffoldTemplateAction.kt:97]: user cancellation again here

File: src/io/confluent/intellijplugin/scaffold/actions/SelectScaffoldTemplateAction.kt [hunk 6]
```diff
@@ -52,15 +83,65 @@ class SelectScaffoldTemplateAction(
...
+                    optionsDialog.optionValues
+                } else {
+                    emptyMap()
```
Comment (Cerchie) [inline src/io/confluent/intellijplugin/scaffold/actions/SelectScaffoldTemplateAction.kt:100]: this means that the project would get generated if you leave all values blank, similar to VS Code

File: src/io/confluent/intellijplugin/scaffold/actions/SelectScaffoldTemplateAction.kt [hunk 7]
```diff
@@ -52,15 +83,65 @@ class SelectScaffoldTemplateAction(
...
+                val chosenDir = fileChooser(project) ?: return@withContext null
+                val templateName = selected.spec.name ?: return@withContext null
+
+                Triple(templateName, options, Path.of(chosenDir.path))
+            } ?: return
+
+            val (templateName, options, targetDir) = userInput
+
+            val hash = UUID.randomUUID().toString().substring(0, 8)
+            val projectDir = targetDir.resolve("${templateName}-${hash}")
```
Comment (Cerchie) [inline src/io/confluent/intellijplugin/scaffold/actions/SelectScaffoldTemplateAction.kt:112]: also how we do it in VS Code

File: src/io/confluent/intellijplugin/scaffold/actions/SelectScaffoldTemplateAction.kt [hunk 8]
```diff
@@ -52,15 +83,65 @@ class SelectScaffoldTemplateAction(
...
+                Files.createDirectories(projectDir)
+                try {
+                    extractZip(zipBytes, projectDir)
+                } catch (ex: Exception) {
+                    projectDir.toFile().deleteRecursively()
+                    throw ex
+                }
+            }
+
+            VfsUtil.markDirtyAndRefresh(true, true, true, projectDir.toFile())
```
Comment (Cerchie) [inline src/io/confluent/intellijplugin/scaffold/actions/SelectScaffoldTemplateAction.kt:132]: confusingly named I think, but this forces IntelliJ to async refresh the file system and reload the children in the dir to IntelliJ before opening the new project

File: src/io/confluent/intellijplugin/scaffold/ui/ScaffoldTemplateSelectionDialog.kt [hunk 9]
```diff
@@ -49,7 +50,6 @@ class ScaffoldTemplateSelectionDialog(
                 comboBox(comboModel, CustomListCellRenderer<ScaffoldV1TemplateListDataInner> {
                     it.spec.displayName ?: it.spec.name ?: KafkaMessagesBundle.message("scaffold.dialog.template.unknown")
                 }).align(AlignX.FILL).resizableColumn().applyToComponent {
-                    templateComboBox = this
                     addActionListener {
                         val selected = comboModel.selectedItem as? ScaffoldV1TemplateListDataInner
```
Comment (Cerchie) [inline src/io/confluent/intellijplugin/scaffold/ui/ScaffoldTemplateSelectionDialog.kt:54]: turns out this line is all we needed to react to the user selection, and didn't need to ref the comboBox directly

File: src/io/confluent/intellijplugin/scaffold/ui/ScaffoldTemplateOptionsDialog.kt [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+                        }
+                        else -> {
+                            textField().align(AlignX.FILL).resizableColumn().applyToComponent {
+                                if (option.initialValue != null) {
+                                    text = option.initialValue
+                                }
+                                if (option.hint != null) {
+                                    putClientProperty("StatusVisibleFunction", null)
+                                    toolTipText = option.hint
+                                    @Suppress("HardCodedStringLiteral")
```
Comment (Cerchie) [inline src/io/confluent/intellijplugin/scaffold/ui/ScaffoldTemplateOptionsDialog.kt:71]: I consider suppressions smelly, here's Copilot's justification of this line, but happy to change approach here.

>Here, the placeholder text is not a hard-coded literal in source code; it comes from option.hint (runtime/template metadata). But the inspection may still be triggered in this context, so the author suppresses it to keep the code warning-free.

File: test/io/confluent/intellijplugin/scaffold/actions/SelectScaffoldTemplateActionTest.kt [hunk 11]
```diff
@@ -336,5 +366,375 @@ class SelectScaffoldTemplateActionTest {
...
+                optionsDialogFactory = optionsDialogFactory
+            )
+
+            runBlocking { action.fetchAndShowTemplates(project) }
+
+            verify(mockOptionsDialog).showAndGet()
+        }
+
+        @Test
+        fun `skips options dialog when template has no options`() {
```
Comment (Cerchie) [inline test/io/confluent/intellijplugin/scaffold/actions/SelectScaffoldTemplateActionTest.kt:406]: i.e., Flink UDFs

File: test/io/confluent/intellijplugin/scaffold/ui/ScaffoldTemplateOptionsDialogTest.kt [hunk 12]
```diff
@@ -0,0 +1,515 @@
...
+            val dialog = createDialog(createTemplate(options))
+
+            onEdt {
+                val combo = dialog.fieldComponents["lang"] as JComboBox<*>
+                assertEquals("Python", combo.selectedItem)
+            }
+        }
+
+        @Test
+        fun `creates no fields for template with no options`() {
```
Comment (Cerchie) [inline test/io/confluent/intellijplugin/scaffold/ui/ScaffoldTemplateOptionsDialogTest.kt:219]: again, as in Flink UDFs related template

**Comment by noeldevelops** [comment]
### Code review

Found 1 issue (minor):

1. Fully qualified type annotation `com.intellij.openapi.Disposable` instead of importing `Disposable` (CLAUDE.md says "Prefer regular Kotlin imports and avoid using fully qualified package names in type annotations")

https://github.com/confluentinc/intellij/blob/98cdb71205c028ac8c0b1d64c6879681c9a59166/test/io/confluent/intellijplugin/scaffold/ui/ScaffoldTemplateOptionsDialogTest.kt#L28-L30

🤖 Generated with [Claude Code](https://claude.ai/code)

<sub>- If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Review by noeldevelops — COMMENTED** [review]
False alarm - error on my end with checkout not up to date :) 
~Click testing locally and for some reason the form never appears after I select a template. Tried several templates and there is no error, just nothing happens after picking one and clicking "OK" (plugin state: clean build, no connection yet)~

File: src/io/confluent/intellijplugin/scaffold/client/ScaffoldHttpClient.kt [hunk 13]
```diff
@@ -130,13 +131,49 @@ class ScaffoldHttpClient(
...
+     *
+     * @param templateName The name of the template to apply
+     * @param collectionName The name of the template collection (default: "intellij")
+     * @param options Key-value options for the template
+     * @return ZIP file bytes
+     * @throws HttpRequests.HttpStatusException if the server returns 4xx or 5xx status
+     */
+    suspend fun applyTemplate(
+        templateName: String,
+        collectionName: String = "intellij",
```
Comment (noeldevelops) [inline src/io/confluent/intellijplugin/scaffold/client/ScaffoldHttpClient.kt:145]: do we expect to use different collections in the future or will it always be intellij here?

File: src/io/confluent/intellijplugin/scaffold/ui/ScaffoldTemplateOptionsDialog.kt (hunk 10 above)
Comment (noeldevelops) [reply src/io/confluent/intellijplugin/scaffold/ui/ScaffoldTemplateOptionsDialog.kt:71]: This is weird. So if you remove this line, we get a warning?

File: src/io/confluent/intellijplugin/scaffold/client/ScaffoldHttpClient.kt (hunk 13 above)
Comment (Cerchie) [reply src/io/confluent/intellijplugin/scaffold/client/ScaffoldHttpClient.kt:145]: the collections map to the interface, so should always be intellij, but the api requires it

**Review by noeldevelops — APPROVED** [review]
This is looking really nice in click-testing! I don't see any issues in the code, let's merge it!

File: src/io/confluent/intellijplugin/scaffold/client/ScaffoldHttpClient.kt (hunk 13 above)
Comment (noeldevelops) [reply src/io/confluent/intellijplugin/scaffold/client/ScaffoldHttpClient.kt:145]: It's a nit so don't worry about it --  imo not necessary to pass as a param, just put in the api path as hard-coded.
